geo/src/algorithm/relate/mod.rs

```diff
@@ -54,13 +54,20 @@ mod relate_operation;
 /// ```
 ///
 /// Note: `Relate` must not be called on geometries containing `NaN` coordinates.
-pub trait Relate<F: GeoFloat> {
-    /// Construct a [`GeometryGraph`]
-    fn geometry_graph(&self, arg_index: usize) -> GeometryGraph<F>;
+pub trait Relate<F: GeoFloat>: BoundingRect<F> + HasDimensions {
+    /// Returns a noded topology graph for the geometry.
+    ///
+    /// # Params
+    ///
+    /// `idx`: 0 or 1, designating A or B (respectively) in the role this geometry plays
+    ///        in the relation. e.g. in `a.relate(b)`
+    fn geometry_graph(&self, idx: usize) -> GeometryGraph<F>;
 
-    fn relate(&self, other: &impl Relate<F>) -> IntersectionMatrix {
-        RelateOperation::new(self.geometry_graph(0), other.geometry_graph(1))
-            .compute_intersection_matrix()
+    fn relate(&self, other: &impl Relate<F>) -> IntersectionMatrix
+    where
+        Self: Sized,
+    {
+        RelateOperation::new(self, other).compute_intersection_matrix()
```

Comment:
This is the crux of the change - rather than passing in the `geometry_graph`, we pass in the thing that `builds` the geometry_graph (which could be a `Geometry` or a `PreparedGeometry`).